pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,188 @@
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		if r.hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node, taintIndex)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if r.hasCniPodRunning(log, pods) {
+			log.V(1).Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node, taintIndex int) error {
+	node.Spec.Taints = slices.Delete(node.Spec.Taints, taintIndex, taintIndex+1)
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func (r *CniNodeTaintReconciler) taintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, kube_core.Taint{
+		Key:    nodeReadinessTaintKey,
+		Effect: kube_core.TaintEffectNoSchedule,
+	})
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("added taint to node")
+	}
+	return err
+}
+
+func (r *CniNodeTaintReconciler) hasCniPodRunning(log logr.Logger, pods []kube_core.Pod) bool {
+	for _, pod := range pods {
+		isCniPod := pod.Labels[cniAppLabel] == r.CniApp
+		if isCniPod && pod.Status.Phase == kube_core.PodRunning {
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == kube_core.PodReady && condition.Status == kube_core.ConditionTrue {
+					log.V(1).Info("pod has kuma-cni-node running and ready", "pod", pod.Name, "status", pod.Status)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func (r *CniNodeTaintReconciler) SetupWithManager(mgr kube_ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &kube_core.Pod{}, nodeIndexField, func(obj kube_client.Object) []string {
+		pod := obj.(*kube_core.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	return kube_ctrl.NewControllerManagedBy(mgr).
+		For(&kube_core.Node{}, builder.WithPredicates(nodeEvents)).
+		Watches(
+			&kube_source.Kind{Type: &kube_core.Pod{}},
+			kube_handler.EnqueueRequestsFromMapFunc(podToNodeMapper(r.Log, r.CniApp)),
+		).
+		Complete(r)
+}
+
+func podToNodeMapper(log logr.Logger, cniApp string) kube_handler.MapFunc {
+	return func(obj kube_client.Object) []kube_reconcile.Request {
+		pod, ok := obj.(*kube_core.Pod)
+		if !ok {
+			log.WithValues("pod", obj.GetName()).Error(errors.Errorf("wrong argument type: expected %T, got %T", pod, obj), "wrong argument type")
+			return nil
+		}
+
+		// For some reason in the logs there are a lot of 'could not find a node with name ""'
+		// so this is why I'm filtering it out here
+		if pod.Spec.NodeName == "" {
```

Comment:
Could this just be because the `Pod` isn't [scheduled yet](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions)? Can we add a predicate on the `Pod` `Watches` for only scheduled pods?

Reply:
Added filtering on the `Watches`

Reply:
@michaelbeaumont even with [this](https://github.com/kumahq/kuma/pull/4650/commits/7990c47efff6e7efec7143e0e5baa7314af35036#diff-68989654770c23c3b103201ae6ff599e37d3745c48b2d679b1c48867fe975931R158-R201) change I'm still seeing one:

```
2022-08-04T10:12:06.466Z        ERROR   controllers.NodeTaint   node not found  {"node": "/", "error": "Node \"\" not found"}
```

in the logs. I have no idea where that event comes from, because I'm using predicates to filter out everything that does not contain node name. Is it worth spending the time on this or just filtering out the way it's filtered out now?

Reply:
Probably because an UpdateEvent triggers two reconciliations, one for the old and one for the new, so for an update where the NodeName is set, the first object won't have it.

Reply:
You're right, I added a check for that and removed the check here.

Reply:
btw it wasn't quite right what I said, an UpdateEvent doesn't trigger two _reconciliations_, it triggers two mapper calls.

Reply:
No, it triggers two reconciliations because I'm seeing `node not found` logs above and they were placed in `Reconcile` function.